username_from_threepid/__init__.py

```diff
@@ -11,36 +11,121 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-
-from typing import Any, Dict
+import re
+import string
+from typing import Any, Dict, Final, Optional
 
 import attr
 from synapse.module_api import ModuleApi
+from synapse.module_api.errors import ConfigError, SynapseError
+
+mxid_localpart_allowed_characters = set(
+    "_-./=" + string.ascii_lowercase + string.digits
+)
+
+
+class LoginType:
+    EMAIL_IDENTITY: Final = "m.login.email.identity"
+    MSISDN: Final = "m.login.msisdn"
 
 
 @attr.s(auto_attribs=True, frozen=True)
-class UsernameFromThreepidConfig:
-    pass
+class UsernameEmailConfig:
```

Comment:
Is `Email` the right name here? The attribute speaks of a threepid, and the `fail_if_not_found` attribute isn't used in an email-specific context.

Reply:
Blugh, no it's not, copy-pasta :(